File: classicexp-mac/ClassicListExporter/DatabaseManager.swift
//
//  DatabaseManager.swift
//  ClassicListExporter
//
//  Created by Casey Cady on 2/3/17.
//  Copyright © 2017 Casey Cady. All rights reserved.
//

import Foundation
import SQLite

struct List {
    var id: Int64?
    var name: String?
    var count: Int64?
}

struct Waypoint {
    var cacheCode: String
    var wptName: String
    var lat: Float
    var lon: Float
    var selected = false
}

class DatabaseManager {
    //let dbFileURL: URL
    let db: Connection
    let token: String
    
    public static var defaultMgr: DatabaseManager?
    
    init?(backupDirectory: URL) {
        let fileMgr = FileManager.default
        
        if let manifestDb = try? Connection(backupDirectory.appendingPathComponent("Manifest.db").absoluteString) {
            do {
                let plistQ = try manifestDb.prepare("select fileId from Files where relativePath = \"Library/Preferences/MZCZ5SMF8U.iCacher.plist\"")
                if let row = plistQ.first(where: { (_) -> Bool in
                    true
                }) {
                    let plistFileId = row[0] as! String
                    var prefix = plistFileId.substring(to: plistFileId.index(plistFileId.startIndex, offsetBy: 2))
                    if let appPlist = NSDictionary(contentsOf: backupDirectory.appendingPathComponent("\(prefix)/\(plistFileId)")) {
                        self.token = appPlist["GEOGlobals_mApiAccessToken"] as! String
                        if let guid = appPlist["GEOGlobals_mLoggedInUserGuid"] as? String {
                            let dbQ = try manifestDb.prepare("select fileId from Files where domain = \"AppDomain-MZCZ5SMF8U.iCacher\" and relativePath = \"Documents/\(guid).db\"")
                            if let dbRow = dbQ.first(where: { (_) -> Bool in
                                true
                            }) {
                                let dbFileId = dbRow[0] as! String
                                prefix = dbFileId.substring(to: dbFileId.index(dbFileId.startIndex, offsetBy: 2))
                                let dbUrl = backupDirectory.appendingPathComponent("\(prefix)/\(dbFileId)")
                                if let db = try? Connection(dbUrl.absoluteString, readonly: true) {
                                    self.db = db
                                    DatabaseManager.defaultMgr = self
                                    return
                                } else {
                                    return nil
                                }
                            }
                        }
                    }
                }
            } catch {
                print(error)
            }
        }
        return nil
    }
    
    
    
    func getCaches(forList list: Int64) -> [String]? {
        let table = Table("geocacheToGroupV3")
        let codeCol = Expression<String>("cacheCode")
        let idCol = Expression<Int64>("groupId")
        let caches = table.select(codeCol).filter(idCol == list)
        return try? db.prepare(caches).flatMap { $0[codeCol] }
    }
    
    func getWaypoints() -> Array<Waypoint> {
        do {
            let r = try db.prepare("SELECT cacheCode, body from userWaypointsV3")
            //return
            let waypoints: [Waypoint] = r.flatMap({
                for (index, name) in r.columnNames.enumerated() {
                    print ("\(name)=\($0[index]!)")
                    // id: Optional(1), email: Optional("alice@mac.com")
                }
                let compressed = Data(base64Encoded: $0[1] as! String)
                let decompressed = compressed?.unzip()
                let dict = try? PropertyListSerialization.propertyList(from: decompressed!, options: [], format: nil) as? [String:Any]
                let records = (dict??["mRecordByAddress"] as! [String:[String:Any]]).values
                
                for record in records {
                    if record["mLatitude"] != nil {
                        return Waypoint(cacheCode: $0[0] as! String, wptName: record["mName"] as! String, lat: record["mLatitude"] as! Float, lon: record["mLongitude"] as! Float, selected: false)
                    }
                }
                return nil
                //return List(id: $0[1] as? Int64, name: $0[0] as? String, count: $0[2] as? Int64)
            })
            return waypoints
            
        } catch {
            print(error)
            return [Waypoint]()
        }
    }
    
    func getLists() -> Array<List>? {
        do {
        let r = try db.prepare("SELECT l.name, l.id, COUNT(c.id) AS cacheCount from geocacheToGroupV3 c left join groupsV3 l on c.groupId = l.id group by l.name")
        return r.flatMap({
            for (index, name) in r.columnNames.enumerated() {
                print ("\(name)=\($0[index]!)")
                // id: Optional(1), email: Optional("alice@mac.com")
            }
            return List(id: $0[1] as? Int64, name: $0[0] as? String, count: $0[2] as? Int64)
        })
        } catch {
            print(error)
            return nil
        }
    }
    
    
}

File: classicexp-win/ClassicListExporter/DatabaseManager.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Data.SQLite;
using System.Data;
using System.IO;
using PList;
using zlib;

namespace ClassicListExporter
{
    public class WaypointItem
    {
        public string name, cacheCode;
        public Double lat, lon;
        public bool IsSelected { get; set; }
        public string waypointDescription
        {
            get { return "Cache :" + cacheCode + " Name: " + name;  }
        }
    }

   public class DatabaseManager
    {
        public string token;
        protected SQLiteConnection conn;
       public DatabaseManager(string backupDir)
        {
            try
            {
                var manifestPath = Path.Combine(backupDir, "Manifest.db");
                var manifestConn = new SQLiteConnection("Data Source=" + manifestPath + ";Version=3;");
                manifestConn.Open();
                DataSet mData = new DataSet();
                string sql = "select fileId from Files where relativePath = \"Library/Preferences/MZCZ5SMF8U.iCacher.plist\"";
                SQLiteCommand command = new SQLiteCommand(sql, manifestConn);
                SQLiteDataAdapter DB = new SQLiteDataAdapter(command.CommandText, manifestConn);
                DB.Fill(mData);
                var fileId = mData.Tables[0].DefaultView[0]["fileId"].ToString();
                var prefix = fileId.Substring(0, 2);
                var infoPath = Path.Combine(backupDir, prefix, fileId);
                //var infoPath = "C:\\Users\\Casey Cady\\Documents\\Documents\\cindy.plist";
                var plist = PListRoot.Load(infoPath).Root as PListDict;
                token = (plist["GEOGlobals_mApiAccessToken"] as PListString).Value;
                var guid = (plist["GEOGlobals_mLoggedInUserGuid"] as PListString).Value;
                sql = "select fileId from Files where domain = \"AppDomain-MZCZ5SMF8U.iCacher\" and relativePath = \"Documents/" + guid + ".db\"";
                command = new SQLiteCommand(sql, manifestConn);
                DB = new SQLiteDataAdapter(command.CommandText, manifestConn);
                mData = new DataSet();
                DB.Fill(mData);
                fileId = mData.Tables[0].DefaultView[0]["fileId"].ToString();
                prefix = fileId.Substring(0, 2);
                var dbPath = Path.Combine(backupDir, prefix, fileId);
                // var dbPath = "C:\\Users\\Casey Cady\\Documents\\Documents\\cindy2.db";
                conn = new SQLiteConnection("Data Source=" + dbPath + ";Version=3;");
                conn.Open();
                manifestConn.Close();
            } catch (Exception e)
            {
                System.Windows.MessageBox.Show(e.ToString());
            }
        }


        ~DatabaseManager()
        {
            //conn.Close();
        }

        public DataSet getLists()
        {
            string sql = "SELECT l.name as name, l.id as id, COUNT(c.id) AS cacheCount from geocacheToGroupV3 c left join groupsV3 l on c.groupId = l.id group by l.name";
            SQLiteCommand command = new SQLiteCommand(sql, conn);
            SQLiteDataAdapter DB = new SQLiteDataAdapter(command.CommandText, conn);
            var listData = new DataSet();
            DB.Fill(listData);
            return listData;
        }

        public DataSet getCaches(string id)
        {
            string sql = "SELECT cacheCode from geocacheToGroupV3 where groupId=\"" + id + "\";";
            SQLiteCommand command = new SQLiteCommand(sql, conn);
            SQLiteDataAdapter DB = new SQLiteDataAdapter(command.CommandText, conn);
            DataSet caches = new DataSet();
            DB.Fill(caches);
            return caches;
        }

        private PListDict decompressPlist(string encodedPList)
        {
            var data = Convert.FromBase64String(encodedPList);
            var z = new ZStream();
            MemoryStream oInStream = new MemoryStream(data);
            MemoryStream oOutStream = new MemoryStream();
            var zIn = new ZInputStream(oInStream);
            byte[] buffer = new byte[2000];
            int len;
            while ((len = zIn.read(buffer, 0, 2000)) > 0)
            {
                oOutStream.Write(buffer, 0, len);
            }
            oOutStream.Flush();
            zIn.Close();

            var plist = PListRoot.Load(oOutStream).Root as PListDict;
            oOutStream.Close();

            return plist;
        }

        public List<WaypointItem> getWaypoints()
        {
            var waypoints = new List<WaypointItem>();
            string sql = "SELECT cacheCode, body from userWaypointsV3;";
            SQLiteCommand command = new SQLiteCommand(sql, conn);
            SQLiteDataAdapter DB = new SQLiteDataAdapter(command.CommandText, conn);
            DataSet wptRows = new DataSet();
            DB.Fill(wptRows);
            foreach (DataRowView row in wptRows.Tables[0].DefaultView)
            {
                var item = new WaypointItem();
                item.cacheCode = row["cacheCode"].ToString();
                var plist = decompressPlist(row["body"].ToString());
                foreach (var plistItem in plist["mRecordByAddress"] as PListDict)
                {
                    var waypointDict = plistItem.Value as PListDict;
                    if (waypointDict.ContainsKey("mLatitude"))
                    {
                        item.name = (waypointDict["mName"] as PListString).Value;
                        item.lat = (waypointDict["mLatitude"] as PListReal).Value;
                        item.lon = (waypointDict["mLongitude"] as PListReal).Value;
                        item.IsSelected = false;
                        waypoints.Add(item);
                        break;
                    }
                }

            }
            return waypoints; 
        }
    }
}
